pkg/controller/azvolumeattachment.go

```diff
@@ -0,0 +1,546 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"sort"
+	"strconv"
+
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/apis/azuredisk/v1alpha1"
+	azVolumeClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type reconcileAzVolumeAttachment struct {
+	client         client.Client
+	azVolumeClient azVolumeClientSet.Interface
+	namespace      string
+	// TODO look for ways to better utilize memory
+	replicaMap map[string]map[string]v1alpha1.AzVolumeAttachment
+	primaryMap map[string]v1alpha1.AzVolumeAttachment
+	nodeMap    map[string]map[string]v1alpha1.AzVolumeAttachment
+	scMap      map[string]map[string]bool
+	volMap     map[string]string
+}
+
+type filteredNode struct {
+	azDriverNode v1alpha1.AzDriverNode
+	numAttached  int
+}
+
+var _ reconcile.Reconciler = &reconcileAzVolumeAttachment{}
+
+func (r *reconcileAzVolumeAttachment) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	if !r.IsInitialized() {
+		if err := r.SyncStorageClass(ctx); err != nil {
+			return reconcile.Result{Requeue: true}, err
+		}
+		if err := r.SyncVolumeAttachments(ctx); err != nil {
+			return reconcile.Result{Requeue: true}, err
+		}
+		klog.Info("Successfully initialized maps for azVolumeAttachment Controller")
+	}
+	// check if request name involves change in storage class
```

Comment:
I do not believe the storage class can be changed on an existing PVC.

Reply:
I reckoned that storage update is normally disallowed but can be done with `--force` flag. Is this something we should not consider? How else would the users dynamically change number of replicas they want for their volumes running on pods? Would they need to create a new storageclass and a new volume to establish that?